poly-commitment/src/srs.rs

```diff
@@ -150,13 +145,75 @@ where
         // because the commitment to the polynomial x^i is just self.g[i], we can obtain
         // commitments to the normalized Lagrange polynomials by applying IFFT to the
         // vector self.g[0..n].
-        let mut lg: Vec<<G as AffineCurve>::Projective> =
-            self.g[0..n].iter().map(|g| g.into_projective()).collect();
-        domain.ifft_in_place(&mut lg);
+        //
+        //
+        // Further still, we can do the same trick for 'chunked' polynomials.
+        //
+        // Recall that a chunked polynomial is some f of degree k*n - 1 with
+        // f(x) = f_0(x) + x^n f_1(x) + ... + x^{(k-1) n} f_{k-1}(x)
+        // where each f_i has degree n-1.
+        //
+        // In the above, if we set u = [ 1, x^2, ... x^{n-1}, 0, 0, .., 0 ]
+        // then we effectively 'zero out' any polynomial terms higher than x^{n-1}, leaving
+        // us with the 'partial Lagrange polynomials' that contribute to f_0.
+        //
+        // Similarly, u = [ 0, 0, ..., 0, 1, x^2, ..., x^{n-1}, 0, 0, ..., 0] with n leading
+        // zeros 'zeroes out' all terms except the 'partial Lagrange polynomials' that
+        // contribute to f_1, and likewise for each f_i.
+        //
+        // By computing each of these, and recollecting the terms as a vector of polynomial
+        // commitments, we obtain a chunked commitment to the L_i polynomials.
+        let srs_size = self.g.len();
+        let num_unshifteds = (n + srs_size - 1) / srs_size;
+        let mut unshifted = Vec::with_capacity(num_unshifteds);
+
+        // For each chunk
+        for i in 0..num_unshifteds {
+            // Initialize the vector with zero curve points
+            let mut lg: Vec<<G as AffineCurve>::Projective> =
+                vec![<G as AffineCurve>::Projective::zero(); n];
+            // Overwrite the terms corresponding to that chunk with the SRS curve points
+            let start_offset = i * srs_size;
+            let num_terms = min((i + 1) * srs_size, n) - start_offset;
+            for j in 0..num_terms {
+                lg[start_offset + j] = self.g[j].into_projective()
+            }
+            // Apply the IFFT
+            domain.ifft_in_place(&mut lg);
+            <G as AffineCurve>::Projective::batch_normalization(lg.as_mut_slice());
+            // Append the 'partial Langrange polynomials' to the vector of unshifted chunks
+            unshifted.push(lg)
+        }
 
-        <G as AffineCurve>::Projective::batch_normalization(lg.as_mut_slice());
-        self.lagrange_bases
-            .insert(n, lg.iter().map(|g| g.into_affine()).collect());
+        // If the srs size does not exactly divide the domain size
+        let shifted: Option<Vec<<G as AffineCurve>::Projective>> =
```

Comment:
why shift the last chunk though?

Reply:
Our chunk size is always `let k = srs.g.len()`. If that doesn't divide the domain size, we want to ensure that the prover hasn't committed to a polynomial larger than the domain. Thus, we need to do an 'upper bound check' by shifting all the way to the right.

Consider a chunked polynomial commitment
```
PolyComm { shifted: Some(s_comm), unshifted: [a_comm, b_comm, c_comm, d_comm] }
```
which corresponds to some evaluations at a point `p` of
```
[a_p, b_p, c_p, d_p]
```

So, we can check using an opening proof that
* `a_comm` opens at `a`,
* `b_comm` opens at `b`,
* `c_comm` opens at `c`,
* `d_comm` opens at `d`

to confirm that the user has committed to some polynomial of maximum degree `4 * k` which evaluates at `p` to `a + p^k * b + p^(2k) * c + p^(3k) * d`. However, if `k` doesn't divide the domain size, the prover's polynomial may be too large, which may allow them to create an invalid proof by cheating the vanishing polynomial division check.

To circumvent this, we make sure that the maximum degree of `d_comm` is `n%k` by providing `s_comm` that opens at `p` to `d * p^{k - n%k}`.

Since `p` is sampled randomly after `d_comm` and `s_comm` are committed to, we know that, with overwhelming probability, for that relationship to hold true we must have
```
d_comm = comm(d(X))
s_comm = comm(X^{k - n%k} * d(X))
```
and thus the prover could only have constructed both commitments if there are commitments to the `deg(d(X)) + k - n%k` powers of `X` available in the srs. Equivalently, `deg(d(X)) + k - n%k <= k`, and so `deg(d(X) <= n%k`.

Putting it all together, we have a polynomial of degree `3*k + n%k = n`, as desired.